Expecta/NSArray+Difference.m

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2020 Lightricks. All rights reserved.
+// Created by Maxim Grabarnik.
+
+#import "NSArray+Difference.h"
+
+#import <assert.h>
+
+#import "ExpectaSupport.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+API_AVAILABLE(ios(13.0))
+typedef NSOrderedCollectionChange<NSObject *> Change;
+
+API_AVAILABLE(ios(13.0))
+@interface ChangePair : NSObject
+
+- (instancetype)init NS_UNAVAILABLE;
+
+- (instancetype)initWithInsertion:(Change *)insertion removal:(Change *)removal;
+
+@property (nonatomic, readonly) Change *insertion;
+
+@property (nonatomic, readonly) Change *removal;
+
+@end
+
+@implementation ChangePair
+
+- (instancetype)initWithInsertion:(Change *)insertion removal:(Change *)removal {
+  if (self = [super init]) {
+    _insertion = insertion;
+    _removal = removal;
+  }
+  return self;
+}
+
+@end
+
+@implementation NSArray (Difference)
+
+- (NSArray<NSString *> *)differenceFrom:(id)object
+    API_AVAILABLE(ios(13.0)) {
+  assert([object isKindOfClass:NSArray.class]);
+  NSArray *expected = self;
+  NSArray *actual = (NSArray *)object;
+
+  NSOrderedCollectionDifference<NSObject *> *diff = [expected differenceFromArray:actual withOptions:NSOrderedCollectionDifferenceCalculationInferMoves];
+
+  NSMutableArray *mismatches = [NSMutableArray array];
+  NSMutableArray *reorders = [NSMutableArray array];
+  NSMutableSet *usedInsertions = [NSMutableSet set];
+  NSMutableSet *usedRemovals = [NSMutableSet set];
+  for (Change *insertion in diff.insertions) {
+    for (Change *removal in diff.removals) {
+      if (insertion.index == removal.index) {
+        [mismatches addObject:[[ChangePair alloc] initWithInsertion:insertion removal:removal]];
+        [usedInsertions addObject:@(insertion.index)];
+        [usedRemovals addObject:@(removal.index)];
+      }
+      if (insertion.associatedIndex == removal.index) {
+        assert(removal.associatedIndex == insertion.index);
+        [reorders addObject:[[ChangePair alloc] initWithInsertion:insertion removal:removal]];
+        [usedInsertions addObject:@(insertion.index)];
+        [usedRemovals addObject:@(removal.index)];
+      }
+    }
+  }
+
+  NSMutableArray *differences = [NSMutableArray array];
+  if (expected.count != actual.count) {
+    [differences addObject:[NSString stringWithFormat:@"Different count: expected (%lu), actual (%lu)",
+                            expected.count, actual.count]];
+  }
+
+  for (ChangePair *mismatch in mismatches) {
+    [differences addObject:[NSString stringWithFormat:@"Mismatch at index (%lu)",
+                            mismatch.insertion.index]];
+    if ([mismatch.insertion.object respondsToSelector:@selector(differenceFrom:)]) {
+      NSArray<NSString *> *diffLines = [(id)mismatch.insertion.object differenceFrom:mismatch.removal.object];
+      NSString *expectedClass = [NSString stringWithFormat:@"%@:", [mismatch.insertion.object class]];
```

Comment:
use `NSStringFromClass()` instead.

Reply:
per discussion added the full type annotation in the following format:
`Expected type SomeType differenceFrom: actual type SomeOtherType:`
_... object diff ..._